mariamihaelafainarea/assignment3/Assignment3

```diff
@@ -0,0 +1,192 @@
+                            if (dictionary.isWord(hopa.toString())) {
+                                theList.add(hopa.toString());
+                            }
+                            DFSonGrid(theList, hopa, newX, newY,visited);
+                        }
+                        //restart the state after a deep search ! 
+                        start.deleteCharAt(hopa.length()-1);
+                        visited[newX][newY] = 0;
+                    }
+                }
+            }
+        }
+        
+    }
+    class MyDictionary{
+        //A dictionary based on a trie ! 
+        //It has a better complexity than
+        //a hashmap !
+        private class TrieNode{
+            private HashMap<Character,TrieNode> children;
+            private boolean endOfWord;
+            //endOfWord is like.. here in this node
+            //a word ends even if it is not a leaf !
+            public TrieNode() {
+                children = new HashMap<>();
+                endOfWord = false;
+            }
+            
+        }
+        private final TrieNode root;
+        
+        public void insert(String word) {
+            //the idea is very simple
+            //you start from root
+            //you go deep down in the trie 
+            //with the current node
+            //if theChildren (the character that need to be added)
+            //exists you only pass it and go deep down
+            TrieNode current = root;
+            for(int i = 0; i < word.length();i++) {
+                TrieNode theChildren = current.children.get(word.charAt(i));
+                if (theChildren == null) {  
+                    theChildren = new TrieNode();
+                    current.children.put(word.charAt(i), theChildren);
+                }
+                current = theChildren;
+            }
+            //after you add the last node and move your
+            //current there
+            //you mark it as endOfWord
+            current.endOfWord = true;
+            
+        }
+        public boolean isWordHelper(TrieNode current,String word) {
+            if (word.equals("")) {
+                return current.endOfWord;
+            }
+            TrieNode theChildren = current.children.get(word.charAt(0));
+            if (theChildren == null) {
+                return false;
+            }
+            return isWordHelper(theChildren, word.substring(1));
+            
+        }
+        public boolean isWord(String word) {
+           //it will call a recursive function that will 
+           //try at each step to find in the children
+           //of the node expanded the character from the word
+           //the auxiliary function get into recursion
+           //with a cut word (shorter) and wait 
+           //the end of the word
+           //to return the value of endOfWord
+           return isWordHelper(root,word);
+        }
+        public TrieNode isPrefixHelper(TrieNode current,String word) {
```

Comment:
The methods isPrefixHelper and isWordHelper are basically completely the same with one small difference. Can you maybe try to merge them smartly so that both isWord and isPrefix can use the same helper?

Reply:
I cannot do this because in prefixHelper it interest me if there is something there. In wordHelper it matters the value of endOfWord. The only posibility i see is to make a generalHelper and to put an if (you search for a prefix) {do this} otherwise {do that}